src/lang/base/Recursion.ml

```diff
@@ -183,12 +183,29 @@ module ScillaRecursion
           let%bind _ = recursion_typ is_adt_in_scope t in
           let%bind new_e = recursion_exp is_adt_in_scope is_adt_ctr_in_scope e in
           pure @@ (x, t, new_e)) cfields in
-    let%bind recursion_ccomps = mapM ~f:(fun c -> recursion_component is_adt_in_scope is_adt_ctr_in_scope c) ccomps in
+    let%bind (recursion_ccomps_rev, _) =
+      foldM ccomps ~init:([], []) ~f:(fun (acc_comps, acc_procs) comp ->
+          let is_proc_in_scope p =
+            match comp.comp_type with
+            | CompTrans ->
+                (* All procedures are in scope. 
```

Comment:
Shouldn't this comment be "All ~procedures~ transitions are in scope. The type checker ensures that the ~procedure~ transition name exists

Reply:
No. We are supposed to check for each procedure call whether the called procedure is in scope (assuming it exists). For procedure calls that occur in transition bodies this is always the case, because all procedures are in scope. However, procedure calls that occur in procedure bodies can only call procedures that have been defined before the current one.

For instance

```
procedure P1(x : ..., y : ...)
  P2 x y
end

procedure P2 (x : ..., y : ...)
...
end
```

is illegal, but

```
transition P1(x : ..., y : ...)
  P2 x y
end

procedure P2 (x : ..., y : ...)
...
end
```

is legal.

It's the same sort of scoping we use for ADTs - the entire library can refer to an ADT, but and ADT definition can only refer to ADTs that have been defined previouly.

Reply:
oh ok, I didn't know that about ADTs either. I thought you can only refer to definitions defined earlier.

For procedures, can't we simply say that all procedures must be defined before all transitions? This is anyway a separate discussion and feel free to merge.